### PR TITLE
fix(dev-infra): run caretaker checks asyncronously

### PR DESCRIPTION
**Pull request by josephperrott**
Run each check in the caretaker check process asyncronously.

### CODE REVIEW

File: dev-infra/caretaker/check/github.ts [hunk 1]
```diff
@@ -64,20 +76,28 @@ async function getGithubInfo(git: GitClient, {githubQueries: queries = []}: Care
...
-        info(`(first ${MAX_RETURNED_ISSUES})`);
-      }
-      for (const node of result.nodes) {
-        info(`- ${node.url}`);
+  const queryResult = await git.github.graphql.query(graphQlQuery);
+  return Object.values(queryResult).map((result, i) => ({
+                                          queryName: queries[i].name,
+                                          count: result.issueCount,
+                                          queryUrl: encodeURI(`https://github.com/${owner}/${
+                                              repo}/issues?q=${queries[i]?.query}`),
```
Comment (gkalpak) [inline dev-infra/caretaker/check/github.ts:84]: Why do we need a `?` in `queries[i]?.query` but not in `queries[i].name` above?
Could we get rid of it here too?

File: dev-infra/caretaker/check/github.ts [hunk 2]
```diff
@@ -64,20 +76,28 @@ async function getGithubInfo(git: GitClient, {githubQueries: queries = []}: Care
...
+                                          matchedUrls: result.nodes.map(node => node.url)
+                                        }));
+}
+
+function printGithubQueryResults(results: GithubQueryResults) {
+  const minQueryNameLength = Math.max(...results.map(result => result.queryName.length));
```
Comment (gkalpak) [inline dev-infra/caretaker/check/github.ts:90]: :100:

**Review by gkalpak — CHANGES_REQUESTED** [review]

**Comment by josephperrott** [comment]
@gkalpak Moved to the "module" system we discussed, PTAL

File: dev-infra/caretaker/check/github.spec.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        queryName: 'key name with spaces',
+        count: 1,
+        queryUrl: 'https://github.com/owner/name/issues?q=issue:%20yes',
+        matchedUrls: ['http://gituhb.com/owner/name/issue/1'],
+      }]);
+    });
+  });
+
+  describe('prints the data retrieved', () => {
+    it('if there are no maches of the query', async () => {
```
Comment (gkalpak) [inline dev-infra/caretaker/check/github.spec.ts:64]:
```suggestion
    it('if there are no matches of the query', async () => {
```

File: dev-infra/caretaker/check/g3.ts [hunk 4]
```diff
@@ -10,76 +10,70 @@ import {existsSync, readFileSync} from 'fs';
...
-  info();
 
+  async printToTerminal() {
+    const stats = await this.data;
+    if (!stats) {
+      return;
+    }
+    info.group(bold('g3 branch check'));
+    if (stats.files === 0) {
+      info(`${stats.commits} commits between g3 and master`);
```
Comment (gkalpak) [inline dev-infra/caretaker/check/g3.ts:46]: OOC, why only log commits when no sync is needed?

**Review by gkalpak — CHANGES_REQUESTED** [review]
Generally lgtm (with a few minor comments/questions/suggestions) :+1:

File: dev-infra/caretaker/check/services.ts [hunk 5]
```diff
@@ -8,72 +8,75 @@
...
-  if (status.status === ServiceStatus.GREEN) {
-    info(`${serviceName} ${green('✅')}`);
-  } else if (status.status === ServiceStatus.RED) {
-    info.group(`${serviceName} ${red('❌')} (Updated: ${status.lastUpdated.toLocaleString()})`);
-    info(`  Details: ${status.description}`);
+  async printToTerminal() {
+    const data = await this.data;
+    const serviceNameMinLength = Math.max(...data.map(service => service.name.length));
+    info.group(bold('Service Statuses'));
+    for (const service of data) {
```
[the remark this answers is not in the document]
Comment (josephperrott) [reply dev-infra/caretaker/check/services.ts:57]: I changed the name of the `data` to `services` which makes it more obvious that we are iterating through the list of  services and looking at each service individually.  Thoughts?

File: dev-infra/caretaker/check/services.spec.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as console from '../../utils/console';
+import {buildVirtualGitClient, mockNgDevConfig, VirtualGitClient} from '../../utils/testing';
+
+import * as servicesExports from './services';
+
+/** Convienence access to ServicesModule symbol. */
+const ServicesModule = servicesExports.ServicesModule;
```
[the remark this answers is not in the document]
Comment (josephperrott) [reply dev-infra/caretaker/check/services.spec.ts:16]: Legacy from when I was having to spy on something exported by `./services`

File: dev-infra/caretaker/check/github.ts [hunk 7]
```diff
@@ -9,75 +9,109 @@
...
 const MAX_RETURNED_ISSUES = 20;
 
-/** Retrieve the number of matching issues for each github query. */
-export async function printGithubTasks(git: GitClient, config?: CaretakerConfig) {
-  if (!config?.githubQueries?.length) {
-    debug('No github queries defined in the configuration, skipping.');
-    return;
+export class GithubQueriesModule extends BaseModule<GithubQueryResults|void> {
+  async retrieveData() {
+    let queries = this.config.caretaker?.githubQueries!;
```
[the remark this answers is not in the document]
Comment (josephperrott) [reply dev-infra/caretaker/check/github.ts:49]: Added comment expressing why its needed.

File: dev-infra/caretaker/check/g3.ts (hunk 4 above)
Comment (josephperrott) [reply dev-infra/caretaker/check/g3.ts:46]: I don't have a good reason, I think it stems from still wanting to express that a difference between the two branches exists, but I have moved it outside the conditional block.

Comment (josephperrott) [reply dev-infra/caretaker/check/g3.ts:46]: Actually, I went back to how it was, but included the information in both resulting statements instead.

File: dev-infra/caretaker/check/services.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply dev-infra/caretaker/check/services.ts]: I still think `status`/`statuses` is more descriptive. For example, `service.description` below sounds as if it refers to the description of the service (while it's the description of the status).

But I don't feel strongly about it - leaving it up to you 👉

[the remark this answers is not in the document]
Comment (josephperrott) [reply dev-infra/caretaker/check/services.ts]: Updated.

File: dev-infra/caretaker/check/ci.spec.ts [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+    await module.printToTerminal();
+
+    expect(debugSpy).toHaveBeenCalledWith('No active release train for name1');
+    expect(infoSpy).toHaveBeenCalledWith('label0 ✅');
+  });
+});
+
+
+/** Build a mock set of ActiveReleaseTrains.  */
```
Comment (gkalpak) [inline dev-infra/caretaker/check/ci.spec.ts:103]:
```suggestion
/** Build a mock set of ActiveReleaseTrains. */
```

😇

**Review by gkalpak — APPROVED** [review]
LGTM 🚀

**Pull request closed without merging**